### ISSUE TITLE
Infinite recursion when generating debuginfo

### ISSUE THREAD

**Issue by tmiasko** [rust-lang/rust#144636]
```rust
use std::marker::PhantomData;

pub struct AB<T: 'static> {
    pub a: PhantomData<T>,
    pub b: *const AB<(T, T)>,
}

fn main() {
    std::hint::black_box(AB::<()> { a: PhantomData, b: std::ptr::null() });
}
```
```console
$ rustc +7278554d82fa474a4e8b5c67afb009e11e41a841 -g a.rs
Terminated
```

It was reintroduced by #144407 to resolve a different issue.

**Comment by theemathas** [comment]
The code compiles fine on the playground with version `1.89.0-beta.7 (2025-07-25 4229c2e708ce796c7174)`, but fails with `1.90.0-nightly (2025-07-29 ba7e63b63871a429533c)`

@rustbot labels +regression-from-stable-to-nightly

**Comment by mati865** [comment]
Cc @godzie44

**Comment by adwinwhite** [comment]
My [overflow check](https://github.com/rust-lang/rust/pull/138599) has false positive, causing the bug in #143241.
```rust
use std::mem::MaybeUninit;
use std::sync::Arc;

struct Inner<T> {
    buffer: Box<MaybeUninit<T>>,
}
struct Shared {
    shared: Arc<Inner<Arc<Shared>>>,
}

fn main() {
    let c = Shared { shared: Arc::new(Inner { buffer: Box::new(MaybeUninit::uninit()) }) };
    std::hint::black_box(c);
}
```
The ADT stack would be(skipping some noise):
- `MaybeUninit<Arc<Shared>>`
- `Arc<Shared>`
- `Shared`
- `Arc<Inner<Arc<Shared>>>`.

With the second `Arc`'s generic arg containing the first one's, the overflow check returns true.

But with cache this code should be fine. It won't make the debuginfo explode.
Let me try to find a fix.

**Comment by adwinwhite** [comment]
simpler,
```rust
struct Wrap<T>(T);

struct Normal {
  a: Wrap<Wrap<Option<Box<Normal>>>>,
}
```
If the debuginfo generation starts with `Wrap<Option<Box<Normal>>>`, then we get the same bug.
- `Wrap<Option<Box<Normal>>>`
- `Option<Box<Normal>>`
- `Box<Normal>`
- `Normal`
- `Wrap<Warp<Option<Box<Normal>>>>`

The edge case I missed is that the generation may start in the middle rather than from the topmost type.

**Comment by apiraino** [comment]
(now in beta)

**Comment by apiraino** [comment]
Assigning priority (discussion on [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/245100-t-compiler.2Fprioritization.2Falerts/topic/.23144636.20Infinite.20recursion.20when.20generating.20debuginfo/near/531745019)).

@rustbot label -I-prioritize +P-medium
